### ISSUE TITLE
[Test] Provide a dummy implementation for virtualMachine object for writing unit tests

### ISSUE THREAD

**Issue by rohanKanojia** [crc-org/crc#4407]
### Component

Internal Testing Utilities

### Description

_Originally discussed in internal chat_

Currently, there are not many unit tests in `pkg/crc/machine` module, probably due to the code dealing directly with managing virtual machines. It makes writing unit tests for CRC start, stop, and delete a bit difficult.

I think we can add some dummy mock implementation for this `virtualMachine` object that can be used while writing unit tests. It can store the state of the `virtualMachine` as some variable that we can expose in tests to assert whether CRC was able to modify the state of VirtualMachine after performing some action (Stop, Start, Delete, etc).

Gerard said they had something like a null driver in minishift codebase that could be used in the tests. Not 100% sure whether this is the correct reference: [minishift@pkg/minishift/provisioner/minishift_provisioner_test.go](https://github.com/minishift/minishift/blob/master/pkg/minishift/provisioner/minishift_provisioner_test.go#L38)

### Acceptance Criteria
- [ ] A dummy implementation is added so that it can be used in tests.
- [ ] code is refactored so that we can this dummy implementation can be picked with some argument (fake driver?).

**Comment by cfergeau** [comment]
Would it be enough to be able to override `loadVirtualMachine` so that it is possible to load a `fakemachine` instead of the real one?

**Comment by rohanKanojia** [comment]
@cfergeau : Yes, that can satisfy my use case. Do you have some examples of how I can achieve that? Perhaps you could share some old pull requests where the team might have added something close to this?

**Comment by cfergeau** [comment]
> Do you have some examples of how I can achieve that? 

@lstocchi did something similar in vfkit unit tests recently, he should be able to help with that.

**Comment by lstocchi** [comment]
@rohanKanojia I joined recently and I didn't have the time to look at the crc project yet so I'm not really able to tell you if what I did is suitable for your use case. You can find the code here https://github.com/crc-org/vfkit/pull/207/commits/41e879f9103d2945ceba40bb0b1861cfa8e6a4c4 . Basically I had to test cloud-init to see if the configuration file was actually read. So I created a fedora machine, mount an image containing the config file, started the vm and interacted with it to verify everything was set accordingly.
You can give that a look to see if it's helpful. 
BTW feel free to ping me if you need help, it would be a chance for me to look at crc 👍

**Comment by cfergeau** [comment]
What I had in mind is the code where you use a function variable to be able to mock a function in a test, I forgot where it is exactly :)

**Comment by rohanKanojia** [comment]
Could someone please assign this issue to me?

**Comment by cfergeau** [comment]
> Could someone please assign this issue to me?

Done!
